src/sentry/incidents/logic.py

```diff
@@ -362,10 +362,6 @@ def bulk_get_incident_event_stats(incidents, query_params_list, data_points=50):
     ]
 
 
-def get_alert_rule_environment_names(alert_rule):
-    return [x.environment.name for x in AlertRuleEnvironment.objects.filter(alert_rule=alert_rule)]
```

Comment:
Would we be removing AlertRuleEnvironment longer term now that QuerySubscriptions also have a connection to environments? Having both seems risky.

Reply:
The problem is that `QuerySubscriptions` needs to be a standalone module, and each `QuerySubscription` should represent a subscription in snuba and be able to recreate itself, so it needs the environments.

With `AlertRules`, we might have more than one `QuerySubscription`, since `AlertRule` is org wide. `AlertRule` needs to know how to create new subscriptions, and while I could just grab a random `QuerySubscription` that is associated with it, that seems risky as well, and so `AlertRules` kind of needs enough information to act as a template for new `QuerySubscriptions` as well. Kind of a pain, but I'm not sure what the better design would be here?

Reply:
If an alert rule has multiple query subscriptions, is it ok for those subscriptions to have different environment lists? If so then my question is moot.

Reply:
It's not, they should have the same environment list. The point though is that I've designed subscriptions to be a different system to alerts, so that we can use subscriptions elsewhere in the codebase without being coupled to alerts. So from the alert rules POV the subscriptions should all have the same environment list, but subscriptions doesn't understand that concept, and if another system uses them it won't necessarily have that restriction.